caddyhttp/proxy/reverseproxy.go

```diff
@@ -126,6 +128,35 @@ func NewSingleHostReverseProxy(target *url.URL, without string) *ReverseProxy {
 	return rp
 }
 
+func NewDynamicHostReverseProxy(templateUrl string, without string) *ReverseProxy {
+	director := func(req *http.Request) {
+		replacer := httpserver.NewReplacer(req, nil, "")
+
+		templateUrl = replacer.Replace(templateUrl)
+
+		target, _ := url.Parse(templateUrl)
+		targetQuery := target.RawQuery
+
+		req.URL.Scheme = target.Scheme
+		req.URL.Host = target.Host
+
+		req.URL.Path = singleJoiningSlash(target.Path, req.URL.Path)
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+
+		// We are then safe to remove the `without` prefix.
+		if without != "" {
+			req.URL.Path = strings.TrimPrefix(req.URL.Path, without)
+		}
+	}
+	rp := &ReverseProxy{Director: director, FlushInterval: 250 * time.Millisecond} // flushing good for streaming & server-sent events
+
+	return rp
+}
```

Comment:
Calling NewSingleHostReverseProxy seem hard as the URL has to change inside the director. The director is also where 90% of the code changes are. What I was thinking of is putting the property setters of the req inside a separate function.